internal/proxy/meta_cache.go

```diff
@@ -607,12 +609,23 @@ func (m *MetaCache) GetShards(ctx context.Context, withCache bool, collectionNam
 		},
 		CollectionID: info.collID,
 	}
-	resp, err := qc.GetShardLeaders(ctx, req)
+
+	// retry until service available or context timeout
+	var resp *querypb.GetShardLeadersResponse
+	childCtx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	err = retry.Do(childCtx, func() error {
+		resp, err = qc.GetShardLeaders(ctx, req)
+		if err != nil {
+			return err
+		}
+		if resp.Status.ErrorCode != commonpb.ErrorCode_Success {
+			return fmt.Errorf("fail to get shard leaders from QueryCoord: %s", resp.Status.Reason)
```

Comment:
Maybe check the error code. Return `retry.Unrecoverable` in some case other than replica not ready.

Reply:
Fixed. Add `NoReplicaAvailable` error code.